esteid/signing/README.md

```diff
@@ -0,0 +1,226 @@
+  If the process is successfully completed, the following methods will be called:
+  * `save_container(self, container, *args, **kwargs)` -- this method receives a temporary BDOC container file handle,
+    and is expected to save it to a persistent store/django model;
+  * `get_success_response(self, *args, **kwargs)` -- is expected to return a response which is meaningful
+    for the application. By default, this just returns a JsonResponse of `{status: "success"}` with a status of 200.
+
+**NOTE:** For security, if signing an existing container, the container is always copied to a temporary file.
+Even though technically the container is only updated with a signature once, after signing is complete,
+we need to insure that we are updating the exact same file that was there at the beginning of the process.
+Also if container is stored in a remote storage, updating it in place can be impossible altogether.
+
+### File types
+
+To generate a signature over files, it is necessary to know the file name, its mime type, and the content.
+To simplify providing these data to `get_files_to_sign()`, a wrapper class `DataFile` is included which
+accepts a path to file, or a django File instance, and a `mime_type` argument, and deals with reading the file content
+when appropriate.
+
+## Implementation details
+
+### Pluggable Signer
+
+The outline of the flow implementation is the Dependency Inversion pattern: 
+instead of the view explicitly calling an implementation, 
+the view interacts with a Signer class, which loads a pluggable implementation
+based on the method selected by user.
+
+```
+POST /path/to/sign/:method
+
+{...params}
+```
+roughly translates to `Signer.start_session(method, session, init_params).prepare(container_path, files)`
+
+```
+GET /path/to/sign/:method
+```
+roughly translates to `Signer.load_session(method, request.session).finalize()`
+
+```
+PATCH /path/to/sign/:method
+
+{...params}
+```
+roughly translates to `Signer.load_session(method, request.session).finalize(params)`
+
+
+### Flow types (backend)
+
+We don't cover the front end part here yet.
+
+* the simplest flow: ID card
+  * the `initial` request
+    - In: the certificate obtained by JS from the browser plugin
+    - Do: prepare the signature container
+    - Out: a digest of the data to sign
+  * the `final` request
+    - In: signed digest (**NOTE:** this requires using the PATCH method to perform the HTTP request)
+    - Do: finalize signature and container (OCSP, TSA)
+    - Out: (can be a redirect to the container download view)
+  * state between requests: the digest, temporary files
+* the external service flow with status: Mobile ID
+  * the `initial` request
+    - In: Personal ID code, Phone number
+    - Do: get user certificate from SK/MID; prepare the container; initiate the `sign` request to SK/MID
+    - Out: signature verification code
+  * the `status/final` request
```

Comment:
I understand it is safe to replay this query until it changes response from `202` to `200`? Is it still safe to replay after it's `200` for second time? This kind of info here could be useful. (And I would probably still use patch always not get, but that is a discussion point, seems to be really an edge-case).

Reply:
Not sure about the meaning of _safe_ here, but as long as this endpoint forwards a status request to the upstream, I wouldn't call it safe. Also once a status has been received and container finalized, we clean up the session, any subsequent calls make no sense and will result in an error. But I'd rather keep it out of this part of the document as it's implementation details.